Objects/memoryobject.c

```diff
@@ -3315,34 +3360,36 @@ memoryiter_dealloc(PyObject *self)
 static int
 memoryiter_traverse(PyObject *self, visitproc visit, void *arg)
 {
-    memoryiterobject *it = (memoryiterobject *)self;
+    memoryiterobject *it = _PyMemoryViewIter_CAST(self);
     Py_VISIT(it->it_seq);
     return 0;
 }
 
+static inline PyObject *
+memoryiter_iter_nth(PyMemoryViewObject *seq, Py_ssize_t nth, const char *fmt)
+{
+    CHECK_RELEASED(seq);
+    const Py_buffer *view = &seq->view;
+    const char *head = (const char *)seq->view.buf;
+    const char *ptr = head + view->strides[0] * nth;
+    ptr = ADJUST_PTR(ptr, view->suboffsets, 0);
+    if (ptr == NULL) {
+        return NULL;
+    }
+    return unpack_single(seq, ptr, fmt);
+}
+
 static PyObject *
 memoryiter_next(PyObject *self)
 {
-    memoryiterobject *it = (memoryiterobject *)self;
-    PyMemoryViewObject *seq;
-    seq = it->it_seq;
+    memoryiterobject *it = _PyMemoryViewIter_CAST(self);
+    PyMemoryViewObject *seq = it->it_seq;
     if (seq == NULL) {
```

Comment:
This needs to raise an exception, otherwise it will raise a `SystemError`. Alternatively, make this an assertion (but make sure that it can *never* be NULL, including in finalizers).

Reply:
Shouldn't it raise a `StopIteration` then?

Reply:
Hmm, I'll look closer at this a little later. If this is called multiple times this will result in a `SystemError` (and assertion failure on debug builds)

Reply:
I believe tp_next conventionally returns NULL without an exception instead of actually raising StopIteration. You can see the same for example in `listiter_next`.

Reply:
Ah, yeah. Apparently, `tp_iternext` can *optionally* raise a `StopIteration`--that seems like an ancient design artifact :(